lib/jxl/dec_frame.cc

```diff
@@ -714,21 +721,53 @@ void FrameDecoder::PreparePipeline() {
     }
   }
 
+  bool late_ec_upsample = frame_header_.upsampling != 1;
+  for (auto ecups : frame_header_.extra_channel_upsampling) {
+    if (ecups != frame_header_.upsampling) {
+      // If patches are applied, either frame_header.upsampling == 1 or
+      // late_ec_upsample is true.
+      late_ec_upsample = false;
+    }
+  }
+
+  if (!late_ec_upsample) {
+    for (size_t ec = 0; ec < frame_header_.extra_channel_upsampling.size();
+         ec++) {
+      if (frame_header_.extra_channel_upsampling[ec] != 1) {
+        builder.AddStage(GetUpsamplingStage(
+            frame_header_.nonserialized_metadata->transform_data, 3 + ec,
+            CeilLog2Nonzero(frame_header_.extra_channel_upsampling[ec])));
+      }
+    }
+  }
+
   if ((frame_header_.flags & FrameHeader::kPatches) != 0) {
     JXL_ABORT("Not implemented: patches");
   }
   if ((frame_header_.flags & FrameHeader::kSplines) != 0) {
     builder.AddStage(
         GetSplineStage(&dec_state_->shared->image_features.splines));
   }
-  // TODO(veluca): extra channels will need some handling too.
+
   if (frame_header_.upsampling != 1) {
     for (size_t c = 0; c < 3; c++) {
       builder.AddStage(GetUpsamplingStage(
           frame_header_.nonserialized_metadata->transform_data, c,
           CeilLog2Nonzero(frame_header_.upsampling)));
     }
   }
+
+  if (late_ec_upsample) {
```

Comment:
Could also integrate it in the loop above, e.g. by having a variable `late_upsample_nb_chans` instead of `late_ec_upsample`, which is equal to 3 in the early ec upsampling case and 3+nb_extra_chans in the other case. In the after-patches upsampling, the upsampling factor is the same anyway...

Reply:
How about this?